PsfLauncher/main.cpp

```diff
@@ -13,223 +13,227 @@
+    if (endScriptInformation)
     {
-        // Non Exe case, use shell launching to pick up local FTA
-        auto nonexePath = packageRoot / exeName;
-
-        CoInitializeEx(NULL, COINIT_APARTMENTTHREADED | COINIT_DISABLE_OLE1DDE);
-        SHELLEXECUTEINFO shex;
-        memset(&shex, 0, sizeof(SHELLEXECUTEINFO));
-        shex.cbSize = sizeof(shex);
-        shex.fMask = SEE_MASK_NOCLOSEPROCESS;
-        shex.hwnd = (HWND)NULL;
-        shex.lpVerb = NULL;  // just use the default action for the fta
-        shex.lpFile = nonexePath.c_str();
-        shex.lpParameters = exeArgString;
-        shex.lpDirectory = dirStr ? (packageRoot / dirStr).c_str() : nullptr;
-        shex.nShow = static_cast<WORD>(cmdShow);
-        
-
-        Log("\tUsing Shell launch: %ls %ls", shex.lpFile, shex.lpParameters);
-        if (!ShellExecuteEx(&shex))
+        ErrorInformation endingScriptError = RunScript(endScriptInformation, packageRoot, dirStr, cmdShow);
+        error.AddExeName(L"Powershell.exe");
+
+        //If there is an existing error from Monitor or the packaged exe
+        if (error.IsThereAnError())
         {
-            std::wostringstream ss;
-            auto err{ ::GetLastError() };
-            // Remove the ".\r\n" that gets added to all messages
-            auto msg = widen(std::system_category().message(err));
-            msg.resize(msg.length() - 3);
-            ss << L"ERROR: Failed to create detoured shell process\n  Path: \"" << exeName << "\"\n  Error: " << msg << " (" << err << ")";
-            ::PSFReportError(ss.str().c_str());
-            return err;
+            ::PSFReportError(error.Print().c_str());
+            return error.GetErrorNumber();
         }
-        else
-        { 
-            // Propagate exit code to caller, in case they care
-            switch (::WaitForSingleObject(shex.hProcess, INFINITE))
-            {
-            case WAIT_OBJECT_0: // SUCCESS
-                if (((unsigned long long)(shex.hInstApp)) > 32)
-                    return 0;
-                else
-                    return (int)(unsigned long long)(shex.hInstApp);
-            case WAIT_FAILED:
-                return ::GetLastError();
-
-            default:
-                return ERROR_INVALID_HANDLE;
-            }			
+        else if (endingScriptError.IsThereAnError())
+        {
+            ::PSFReportError(endingScriptError.Print().c_str());
+            return endingScriptError.GetErrorNumber();
         }
     }
+
+    return 0;
 }
 catch (...)
 {
     ::PSFReportError(widen(message_from_caught_exception()).c_str());
     return win32_from_caught_exception();
+}
 
+ErrorInformation RunScript(const psf::json_object * scriptInformation, std::filesystem::path packageRoot, LPCWSTR dirStr, int cmdShow)
+{
+    //Script arguments are optional.
+    auto scriptArgumentsJObject = scriptInformation->try_get("scriptArguments");
+    std::wstring scriptArguments;
+    if (scriptArgumentsJObject)

```

Comment:
not needed - just move the check below to where you append

Reply:
How would I do that?  A long ternary if statement?

I understand where you want me to put it.  Where I do the appending, but, I need to check to see if the json_oibject exists before I use it.  So, I need to first grab the json_object, then check to see if it exists.  I don't know how I would reduce it.

Reply:
auto scriptArgumentsJObject = scriptInformation->try_get("scriptArguments");
    if (scriptArgumentsJObject)
{
powershellCommandString.append(scriptArgumentsJObject->as_string().wide(););
}